Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,293 @@
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import TSCBasic
+import TSCUtility
+
+typealias TSCDiagnostic = TSCBasic.Diagnostic
+
+// this could become a struct when we remove the "errorsReported" pattern
+public class ObservabilitySystem {
+
+    // global
+
+    private static var _global = ObservabilitySystem(factory: NOOPFactory())
+    private static var bootstrapped = false
+    private static let lock = Lock()
+
+    public static func bootstrapGlobal(factory: ObservabilityFactory) {
+        Self.lock.withLock {
+            // FIXME: disabled for testing
+            //precondition(!Self.bootstrapped, "ObservabilitySystem can only bootstrapped once")
+            Self._global = .init(factory: factory)
+            Self.bootstrapped = true
+        }
+    }
+
+    @available(*, deprecated, message: "this pattern is deprecated, transition to error handling instead")
+    public static var errorsReported: Bool {
+        Self.lock.withLock {
+            Self._global.errorsReported
+        }
+    }
+
+    // as we transition to async/await we can take advantage of Task Local values instead of a global
+    fileprivate static var global: ObservabilitySystem {
+        Self.lock.withLock {
+            Self._global
+        }
+    }
+
+    // compatibility with DiagnosticsEngine
+
+    @available(*, deprecated, message: "temporary for transition DiagnosticsEngine -> DiagnosticsEmitter")
+    public static func makeDiagnosticsEngine() -> DiagnosticsEngine {
+        return DiagnosticsEngine(handlers: [{ Diagnostic($0).map{ self.global.diagnosticsHandler($0) } }])
+    }
+
+    // instance
+
+    public var diagnosticsHandler: DiagnosticsHandler!
+    private var _errorsReported = ThreadSafeBox<Bool>(false)
+
+    public init(factory: ObservabilityFactory) {
+        self.diagnosticsHandler = { diagnostic in
+            if diagnostic.severity == .error {
+                self._errorsReported.put(true)
+            }
+            factory.diagnosticsHandler(diagnostic)
+        }
+    }
+
+    // FIXME: we want to remove this functionality and move to more conventional error handling
+    @available(*, deprecated, message: "this pattern is deprecated, transition to error handling instead")
+    public var errorsReported: Bool {
+        self._errorsReported.get() ?? false
+    }
+
+    private struct NOOPFactory: ObservabilityFactory {
+        var diagnosticsHandler: DiagnosticsHandler = { _ in }
+    }
+}
+
+public protocol ObservabilityFactory {
+    var diagnosticsHandler: DiagnosticsHandler { get }
+}
+
+public typealias DiagnosticsHandler = (Diagnostic) -> Void
+
+public struct Diagnostic: CustomStringConvertible, Equatable {
+    public let context: DiagnosticsContext?
+    private let underlying: DiagnosticMessage
```

Comment:
Is this still here just because of TSCDiagnostic?  In general I've always disliked the extra complexity of having both a diagnostic and a message in that system.  I wonder if it's possible to combine them?

Reply:
was following the existing example, but I agree we can flatten and that would simplify things